vowpalwabbit/namespaced_features.h

```diff
@@ -0,0 +1,201 @@
+  std::pair<indexed_iterator, indexed_iterator> get_namespace_index_groups(namespace_index index);
+  // Returns empty range if not found
+  std::pair<const_indexed_iterator, const_indexed_iterator> get_namespace_index_groups(namespace_index index) const;
+
+  // If a feature group already exists in this "slot" it will be merged
+  template <typename FeaturesT>
+  features& merge_feature_group(FeaturesT&& ftrs, uint64_t hash, namespace_index index);
+
+  // If no feature group already exists a default one will be created.
+  // Creating new feature groups will invalidate any pointers or references held.
+  features& get_or_create_feature_group(uint64_t hash, namespace_index index);
+
+  const features& operator[](uint64_t hash) const;
+  features& operator[](uint64_t hash);
+
+  // Removing a feature group will invalidate any pointers or references held.
+  void remove_feature_group(uint64_t hash);
+
+  generic_range<indexed_iterator> namespace_index_range(namespace_index index);
+  generic_range<const_indexed_iterator> namespace_index_range(namespace_index index) const;
+  indexed_iterator namespace_index_begin(namespace_index index);
+  indexed_iterator namespace_index_end(namespace_index index);
+  const_indexed_iterator namespace_index_begin(namespace_index index) const;
+  const_indexed_iterator namespace_index_end(namespace_index index) const;
+  const_indexed_iterator namespace_index_cbegin(namespace_index index) const;
+  const_indexed_iterator namespace_index_cend(namespace_index index) const;
+
+  iterator begin();
+  iterator end();
+  const_iterator begin() const;
+  const_iterator end() const;
+  const_iterator cbegin() const;
+  const_iterator cend() const;
+
+private:
+  std::vector<features> _feature_groups;
+  std::vector<namespace_index> _namespace_indices;
+  std::vector<uint64_t> _namespace_hashes;
+
+  std::unordered_map<namespace_index, std::vector<size_t>> _legacy_indices_to_index_mapping;
+  std::unordered_map<uint64_t, size_t> _hash_to_index_mapping;
+};
+
+// If a feature group already exists in this "slot" it will be merged
+// Creating new feature groups will invalidate any pointers or references held.
+template <typename FeaturesT>
+features& namespaced_features::merge_feature_group(FeaturesT&& ftrs, uint64_t hash, namespace_index ns_index)
+{
+  auto* existing_group = get_feature_group(hash);
+  if (existing_group == nullptr)
+  {
+    _feature_groups.emplace_back(std::forward<FeaturesT>(ftrs));
+    _namespace_indices.push_back(ns_index);
+    _namespace_hashes.push_back(hash);
+    auto new_index = _feature_groups.size() - 1;
+    _hash_to_index_mapping[hash] = new_index;
+    _legacy_indices_to_index_mapping[ns_index].push_back(new_index);
+    existing_group = &_feature_groups.back();
+  }
+  else
+  {
+    existing_group->concat(std::forward<FeaturesT>(ftrs));
+    auto existing_index = _hash_to_index_mapping[hash];
+    // Should we ensure that this doesnt already exist under a DIFFERENT namespace_index?
```

Comment:
maybe add assertion for this comment?

Reply:
This would require either scanning all mappings or a reverse lookup which is why I didnt add an assertion